woocommerce-gateway-stripe.php

```diff
@@ -593,16 +599,44 @@ public function register_routes() {
 					$upe_flag_toggle_controller = new WC_Stripe_REST_UPE_Flag_Toggle_Controller();
 					$upe_flag_toggle_controller->register_routes();
 
-					$gateway = WC()->payment_gateways()->payment_gateways()[ WC_Gateway_Stripe::ID ];
-
 					require_once WC_STRIPE_PLUGIN_PATH . '/includes/admin/class-wc-rest-stripe-settings-controller.php';
-					$settings_controller = new WC_REST_Stripe_Settings_Controller( $gateway );
+					$settings_controller = new WC_REST_Stripe_Settings_Controller( $this->get_main_stripe_gateway() );
 					$settings_controller->register_routes();
 
 					$stripe_account_keys_controller = new WC_REST_Stripe_Account_keys_Controller();
 					$stripe_account_keys_controller->register_routes();
 				}
 			}
+
+			/**
+			 * Returns the main Stripe payment gateway class instance.
+			 *
+			 * @return WC_Stripe_Payment_Gateway
+			 */
+			public function get_main_stripe_gateway() {
+				if ( ! is_null( $this->stripe_gateway ) ) {
```

Comment:
I don't think this one property is enough to:
>  only one instance of the Stripe gateway is always present.

The core code will construct it's own instance of the gateway which is why we were using `WC()->payment_gateways()->payment_gateways()[ WC_Gateway_Stripe::ID ]`
We should try this core method first and if the gateway is not set there create our own instance.

As the code is multiple instances could exist registering duplicate filters in their constructor.

Reply:
The core code won't construct its own instance, because in the `woocommerce_payment_gateways` filter we're passing the instance from the `get_main_stripe_gateway()` method.
When `woocommerce_payment_gateways` is called in core, the code already checks if it's a class or if it's a string representing a class: https://github.com/woocommerce/woocommerce/blob/0c44ab857a9b041727ab8c16fd182ee9b700818e/includes/class-wc-payment-gateways.php#L95-L113

Let me know if I missed something there!

Reply:
My bad. I missed that you are passing an object [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2023/files#diff-e1e5e9b442c2b5d4105edbdc3787651333c1923dd623677025f72d25796900d6R380) now.